Message interviews:18 -  Read
From:      Paul Calder <calder@aramis.stanford.edu>
To:        <interviews@interviews.stanford.edu>
Reply-To:  <calder@lurch.stanford.edu>
Subject:   InterViews and stipple problems on Xqdss servers
Obsoletes: Sanjay Ghemawat's message of Fri, 30 Jun 89 23:00:26 -0400

Sanjay reports a (known) bug in the Xqdss server that is tickled by
InterViews 2.5.  His mail also points out an oversight in the use of
Patterns in InterViews.  Briefly, InterViews tries to optimise the
common case of area filling with a solid color, which can be done more
cheaply than filling with a stipple pattern.  However this
optimization is currently not done correctly and stipple filling is
always used.

Sanjay suggests a fix based on comparing the fill pattern with the
predefined pattern 'solid'.  Instead, the following fix will work for
a wider range of "solid" patterns (and is the way it was supposed to
work!)

This patch is also available for ftp as 'fix9'
--------------------------------------------------------
*** X11.c.old   Sat Jul  1 11:30:31 1989
--- X11.c       Sat Jul  1 11:32:03 1989
***************
*** 1891,1915 ****
  }

  Pattern::Pattern (int dither) {
!     register int i, seed;
!     int r[16];

!     seed = dither;
!     for (i = 0; i < 4; i++) {
!       r[i] = (seed & 0xf000) >> 12;
!       r[i] |= r[i] << 4;
!       r[i] |= r[i] << 8;
!       seed <<= 4;
!       r[i+4] = r[i];
!       r[i+8] = r[i];
!       r[i+12] = r[i];
      }
-     info = (void*)MakeStipple(r);
  }

  Pattern::~Pattern () {
      if (LastRef()) {
!       XFreePixmap(_world->display(), (Pixmap)info);
      }
  }

--- 1891,1921 ----
  }

  Pattern::Pattern (int dither) {
!     if (dither == 0xffff) {
!         info = nil;
!     } else {
!         register int i, seed;
!         int r[16];

!         seed = dither;
!         for (i = 0; i < 4; i++) {
!             r[i] = (seed & 0xf000) >> 12;
!             r[i] |= r[i] << 4;
!             r[i] |= r[i] << 8;
!             seed <<= 4;
!             r[i+4] = r[i];
!             r[i+8] = r[i];
!             r[i+12] = r[i];
!         }
!         info = (void*)MakeStipple(r);
      }
  }

  Pattern::~Pattern () {
      if (LastRef()) {
!         if (info != nil) {
!             XFreePixmap(_world->display(), (Pixmap)info);
!         }
      }
  }
